### PR TITLE
feat:[#272] add flatpak packaging

### LINKED ISSUE #272: [SUGGESTION] Build as Flatpak
Building as a Flatpak would be great for Linux users.
Also, it would normalize the behavior on lal devices and facilitate installation on steam deck.

### PR DESCRIPTION
**Pull request by jardon**
- add flatpak manifest
- add build dirs to gitignore
- add flatpak build workflow job
- add update support for flatpak package

closes #272

### CODE REVIEW

**Comment by ngoquang2708** [comment]
Do you plan to publish to Flathub or just to build a flatpak bundle?

**Comment by jardon** [comment]
i dont plan on anything specifically.  personally, i think it should be distributed on flathub, but thats for someone on the team to decide and put together.

**Comment by jardon** [comment]
i can also put together github actions for building this, but i couldnt get a straight answer on discord so i left it out.

**Comment by ngoquang2708** [comment]
Since ```rapidjson``` and ```libfmt``` modules are static linked. I think you should include a clean step for each of them.
```
cleanup:
  - '*'
```

**Comment by ngoquang2708** [comment]
Also translations is not included in the flatpak.

**Comment by ngoquang2708** [comment]
While you are at it, fix the emulator names in the .desktop file to the correct one: ```shadPS4```.

**Comment by ngoquang2708** [comment]
~~Another issue is that the emulator don't remember game install dir if that dir is not in the user home.~~

**Comment by ngoquang2708** [comment]
The emulator is shown as "Emulator" in Flatseal.
Screenshot_20241104_112035

**Comment by ngoquang2708** [comment]
Some games don't have music and voice dialog. This fixed it for me:
```
add-extensions:
  org.freedesktop.Platform.ffmpeg-full:
    directory: lib/ffmpeg
    add-ld-path: .
    version: '24.08'
    autodownload: true
    autodelete: false
cleanup-commands:
  - mkdir -p ${FLATPAK_DEST}/lib/ffmpeg
```

File: com.shadps4.Emulator.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+command: shadps4
+rename-icon: shadps4
+rename-desktop-file: shadps4.desktop
+finish-args:
+  - --share=network
+  - --socket=pulseaudio
+  - --socket=wayland
+  - --socket=fallback-x11
+  - --allow=bluetooth
+  - --filesystem=home
```
Comment (ngoquang2708) [inline com.shadps4.Emulator.yaml]: Consider add this:
```
  - --filesystem=/media
  - --filesystem=/run/media
```

Comment (ngoquang2708) [reply com.shadps4.Emulator.yaml]: This make loading gamelist must faster on external drives.

**Comment by ngoquang2708** [comment]
I cannot figure out how to enable MangoHud. I don't know if this is our issue.

**Comment by ngoquang2708** [comment]
On KDE Plasma 6.7, there is no mute icon in the task manager: without-mute-icon
Non-flatpak builds do have: with-mute-icon

**Comment by ngoquang2708** [comment]
> While you are at it, fix the emulator names in the .desktop file to the correct one: `shadPS4`.

I think this will break the name of the output artifact of appimages thus scripts will fail.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -382,11 +382,32 @@ jobs:
...
+    - uses: flatpak/flatpak-github-actions/flatpak-builder@master
+      with:
+        manifest-path: com.shadps4emu.ShadPS4.yaml
+        cache-key: flatpak-builder-${{ github.sha }}
+        bundle: shadps4.flatpak
+        upload-artifact: false
+    - uses: actions/upload-artifact@v4
+      with:
+        path: shadps4.flatpak
+        name: shadps4-linux-${{ needs.get-info.outputs.date }}-${{ needs.get-info.outputs.shorthash }}.flatpak
```
Comment (ngoquang2708) [inline .github/workflows/build.yml]: ~~It seems that the ```upload-artifact``` action will automatically compress the artifact.~~ These should be changed to match the others.
Screenshot_20241104_214355

```suggestion
        name: shadps4-linux-flatpak-${{ needs.get-info.outputs.date }}-${{ needs.get-info.outputs.shorthash }}
```

File: com.shadps4emu.ShadPS4.yaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# SPDX-FileCopyrightText: Copyright 2024 shadPS4 Emulator Project
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+id: com.shadps4emu.ShadPS4
```
Comment (ngoquang2708) [inline com.shadps4emu.ShadPS4.yaml]: We have a website: https://shadps4.net/.
Plus, should we make shadPS4 name more consistent?
```suggestion
id: net.shadps4.shadPS4
```

Comment (ngoquang2708) [reply com.shadps4emu.ShadPS4.yaml]: Manifest file name too:
```
com.shadps4emu.ShadPS4.yaml
```
->
```
net.shadps4.shadPS4.yaml
```

**Comment by ngoquang2708** [comment]
We have the mute icon now.

**Comment by ngoquang2708** [comment]
Look good in Flatseal now:
Screenshot_20241104_221421

**Comment by ngoquang2708** [comment]
Seem we are missing a ```metainfo.xml``` entry/file:
```
rename-appdata-file: shadps4.metainfo.xml
```
so KDE Plasma Settings don't display the correct app name:

Screenshot_20241104_222057

**Comment by qurious-pixel** [comment]
> I cannot figure out how to enable MangoHud. I don't know if this is our issue.

This requires the Mangohud flatpak.

**Comment by jardon** [comment]
> Also translations is not included in the flatpak.

afaik [this](https://github.com/shadps4-emu/shadPS4/blob/76f440889b28b01a32dc47b427e85a13bfb8e32d/CMakeLists.txt#L764) should include the translations as long as the QT GUI is enabled.

Can anyone confirm?

**Comment by jardon** [comment]
@dima-xd could you provide any insight here?

**Comment by dima-xd** [comment]
> @dima-xd could you provide any insight here?

Yeah, you're right. Translations should work as they are

**Comment by jardon** [comment]
ive never dealt with cmake translations, but the translations directory and the `.qm` and the `.qrc` file shouldnt be needed after compilation afaik

**Comment by dima-xd** [comment]
> ive never dealt with cmake translations, but the translations directory and the `.qm` and the `.qrc` file shouldnt be needed after compilation afaik

Yep, it was a mistake from me to add them. So it was never fixed

**Comment by ngoquang2708** [comment]
There are 4 places we've missed to change the emulator name in the embedded update script here https://github.com/shadps4-emu/shadPS4/blob/76f440889b28b01a32dc47b427e85a13bfb8e32d/src/qt_gui/check_update.cpp#L471-L481

**Comment by ngoquang2708** [comment]
> > I cannot figure out how to enable MangoHud. I don't know if this is our issue.
> 
> This requires the Mangohud flatpak.

@qurious-pixel
Ah, it works. It just doesn't response to hotkey. I have to change the config to make it to always show.
Now I need to figure out why it doesn't response to hotkey. Any ideas?

**Comment by Xxp0mxX** [comment]
tested on manajaro kde, installed using terminal, no issues with the flatpak itself so far

**Comment by georgemoralis** [comment]
if it is ok you can remove the draft so we can merge

**Comment by jardon** [comment]
> if it is ok you can remove the draft so we can merge

@georgemoralis I would rather fix the updater first if that is okay with you

File: src/qt_gui/check_update.cpp [hunk 4]
```diff
@@ -462,23 +473,51 @@ void CheckUpdate::Install() {
...
+        "        chmod +x \"$2/shadPS4-qt.AppImage\"\n"
+        "   else\n"
+        "        flatpak install -u \"$2/shadps4.flatpak\"\n"
+        "   fi\n"
+        "}\n"
+        "restart_app() {\n"
+        "   if [ \"$TYPE\" == \"appimage\" ]; then\n"
+        "        cd \"$1\" && ./shadPS4-qt.AppImage\n"
+        "   else\n"
+        "        flatpak run -u com.shadps4emu.ShadPS4\n"
```
Comment (ngoquang2708) [inline src/qt_gui/check_update.cpp]:
```suggestion
        "        flatpak run -u net.shadps4.shadPS4\n"
```

File: src/qt_gui/check_update.cpp [hunk 5]
```diff
@@ -462,23 +473,51 @@ void CheckUpdate::Install() {
...
         "    sleep 2\n"
         "    extract_file\n"
         "    sleep 2\n"
-        "    if pgrep -f \"Shadps4-qt.AppImage\" > /dev/null; then\n"
-        "        pkill -f \"Shadps4-qt.AppImage\"\n"
-        "        sleep 2\n"
-        "    fi\n"
-        "    cp -r \"%2/\"* \"%3/\"\n"
+        "    set_package_type\n"
+        "    sleep 2\n"
```
Comment (ngoquang2708) [inline src/qt_gui/check_update.cpp:507]:
```suggestion
        "    sleep 2\n"
        "    kill_process\n"
        "    sleep 2\n"
```

File: src/qt_gui/check_update.cpp [hunk 6]
```diff
@@ -462,23 +466,53 @@ void CheckUpdate::Install() {
...
+        "        pkill -f \"Shadps4\"\n"
+        "        sleep 2\n"
+        "    fi\n"
+        "}\n"
+        "update_package() {\n"
+        "   if [ \"$TYPE\" == \"appimage\" ]; then\n"
+        "        cp -r \"$1/\"* \"$2/\"\n"
+        "        chmod +x \"$2/Shadps4-qt.AppImage\"\n"
+        "   else\n"
+        "        flatpak install -u \"$2/shadps4.flatpak\"\n"
```
Comment (ngoquang2708) [inline src/qt_gui/check_update.cpp:490]: Should be this?
```suggestion
        "        flatpak install -u --noninteractive --assumeyes \"$1/shadps4.flatpak\"\n"
```

Comment (ngoquang2708) [reply src/qt_gui/check_update.cpp:490]: But this won't work inside of the container.

**Comment by jardon** [comment]
closing in favor of the out-of-tree approach for easier flathub support.

**Pull request closed without merging**